src/geohash_helper.c

```diff
@@ -99,11 +99,55 @@ uint8_t geohashEstimateStepsByRadius(double range_meters, double lat) {
  */
 int geohashBoundingBox(GeoShape *shape, double *bounds) {
     if (!bounds) return 0;
+    double height = 0.0, width = 0.0;
+    if (shape->type == CIRCULAR_TYPE) {
+        height = shape->conversion * shape->t.radius;
+        width = shape->conversion * shape->t.radius;
+    } else if (shape->type == RECTANGLE_TYPE) {
+        height = shape->conversion * shape->t.r.height / 2;
+        width = shape->conversion * shape->t.r.width / 2;
+    } else if (shape->type == POLYGON_TYPE) {
+        int num_vertices = shape->t.polygon.num_vertices;
+        double x = 0.0, y = 0.0, z = 0.0;
+        /* Bounding box directly from lon & lat. */
+        double min_lon = GEO_LONG_MAX, max_lon = GEO_LONG_MIN;
+        double min_lat = GEO_LAT_MAX, max_lat = GEO_LAT_MIN;
+        for (int i = 0; i < num_vertices; i++) {
+            double longitude = shape->t.polygon.points[i][0];
+            double latitude = shape->t.polygon.points[i][1];
+            /* Update bounding box (in degrees). */
+            if (longitude < min_lon) min_lon = longitude;
+            if (longitude > max_lon) max_lon = longitude;
+            if (latitude < min_lat) min_lat = latitude;
+            if (latitude > max_lat) max_lat = latitude;
+            /* Convert to Cartesian coordinates and accumulate for centroid. */
+            double lon_rad = deg_rad(longitude);
+            double lat_rad = deg_rad(latitude);
+            double cur_x = EARTH_RADIUS_IN_METERS * cos(lat_rad) * cos(lon_rad);
+            double cur_y = EARTH_RADIUS_IN_METERS * cos(lat_rad) * sin(lon_rad);
+            double cur_z = EARTH_RADIUS_IN_METERS * sin(lat_rad);
+            x += cur_x;
+            y += cur_y;
+            z += cur_z;
```

Comment:
I don't really understand this. If I don't then I guess code readers also don't so some code comment to explain this to the readers would be good.

Do I understand correctly:

1. Long and lat are in degrees.
2. Convert to radians so we can use cos and sin on them.
3. Multiply with the radius of the earth...

I don't fully understand this Cartesian coordinate system. It starts from the center of the earth. The Z axis is north-south, the X axis is towards somewhere south of Greenwich (London) and the Y axis is the Equador-Bangladesh axis, 90 degrees from the Greenwich?

The point (x,y,z) is on the surface, right? Is the average of all such points is still on the surface? Or that doesn't matter?

Reply:
> Do I understand correctly:

Long and lat are in degrees.
Convert to radians so we can use cos and sin on them.

This is correct

> Multiply with the radius of the earth...

If I think about this more, we do not need to multiply with the radius here.
The centroid calculation can be done just by summing the Cartesian coordinates. The radius is needed only when calculating the distance between points on earth (this is why we use it in `geohashGetDistance`).

Reply:
So no need to divide the sum by the number of points either then?

Reply:
We will have to sum up the Cartesian coordinates, divide by num vertices to get the average and then convert back to LON/LAT since other following operations for the Geo Hash based KNN search depend on the `shape` containing LON / LAT. 

> divide by num vertices to get the average and then convert back to LON/LAT

Without divi, the result will not represent the actual centroid's position on the sphere, but instead it will be a vector sum.

Reply:
If we don't divide by the number of verices, the sum will be a point high up in space, but the same angle (long, lat) as if we divide, isn't it? Muliplying and dividing just moves the point higher or lower (towards or away from earth center).

So we can skip multiply by earth radius and divide by num vertices and still have the same angle. We just need the trigonometry to go back to long-lat. No?

Reply:
> If we don't divide by the number of verices, the sum will be a point high up in space, but the same angle (long, lat) as if we divide, isn't it? 

Yes, I see what you mean. You are definitely right. We don't need to worry about the magnitude here. The division is just for more correctness in the "average"

But the angle is all that matters, so we can skip the division to save on some computations